src/porepy/viz/exporter.py

```diff
@@ -180,6 +182,183 @@ def __init__(
+                if "are_subdomain_data" not in kwargs:
+                    is_subdomain_data = True
+                else:
+                    are_subdomain_data = kwargs.pop("are_subdomain_data")
+                    assert isinstance(are_subdomain_data, bool) or isinstance(
+                        are_subdomain_data, list
+                    )
+                    is_subdomain_data = (
+                        are_subdomain_data[i]
+                        if isinstance(are_subdomain_data, list)
+                        else are_subdomain_data
+                    )
+
+            # 2. Step: Make sure that the vtu file and the current grid are identical.
+            # For this, check whether the meshio geometry is the same as in the vtu file.
+            # NOTE: Meshes could be compatible via some transformation. However,
+            # here we require identical grids.
+
+            meshio_geometry = (
+                self.meshio_geom[dim] if is_subdomain_data else self.m_meshio_geom[dim]
+            )
+            assert isinstance(meshio_geometry, Meshio_Geom)
+
+            # Make sure that both grids have same nodes
+            assert np.all(np.isclose(meshio_geometry.pts, vtu_data.points))
+
+            for i, connectivity in enumerate(meshio_geometry.connectivity):
+
+                # Make sure that the connectivity patterns are identical
+                assert (
+                    DeepDiff(
+                        connectivity,
+                        vtu_data.cells[i],
+                        significant_digits=8,
+                        number_format_notation="e",
+                        ignore_numeric_type_changes=True,
+                    )
+                    == {}
+                )
+
+            # 3. Step: Transfer data. Consider each key separately.
+            for key in keys:
+
+                # Only continue if the key is present in the data
+                if key in vtu_data.cell_data:
+
+                    # Data is stored in a list with each element storing data for one specific
+                    # cell type (most relevant for polygonal and polyhedral grids). Accumulate
+                    # the data again, assuming the same dimensionality in each cell.
+                    value = np.concatenate(tuple(vtu_data.cell_data[key]), axis=0)
+
+                    # Chop data in pieces compatible with the subdomains and interfaces
+                    offset = 0
+                    if is_subdomain_data:
+                        for sd, sd_data in self._mdg.subdomains(
+                            dim=dim, return_data=True
+                        ):
+                            if pp.STATE not in sd_data:
+                                sd_data[pp.STATE] = {}
+                            sd_data[pp.STATE][key] = _toScalarFormat(
+                                value[offset : offset + sd.num_cells], sd
+                            )
+                            offset += sd.num_cells
```

Comment:
This assumes that the ordering of the state in the current `mdg` is the same as that used to generate the vtu file. I guess that will always be true, but it will also be incredibly hard to detect if this is not the case. I'm not sure we can do anything about it, though.

Reply:
What do you mean by "ordering of the state"? Ordering of subdomains? If so, proper subdomain tags could go some way.

Reply:
My impression is that @keileg thinks of the ordering of subdomains. Yes, some unique sorting of subdomains would help to lower the risk. The hope is the second step of the import already lowers the risk in some parts, by checking whether the grids comply to each other. For non-articifical scenarios (for which importing becomes more interesting) it would be a miracle that the multiple subdomains have the same number of cells e.g. I believe the second step would catch this error already.

Reply:
Note that the compatibility of the vtu grid and the considered `MixedDimensionalGrid` is checked. It is not a full guarantee, but one should be very unlucky if the compatibility succeeds by the ordering is different.